### PR TITLE
Sb track2 send Via feature - Sending message to a Queue via another queue 

### PR DESCRIPTION
**Pull request by hemanttanwar**
- Added send via feature
     - create Sender Link with appropriate link properties
-- This happens in a transaction

### CODE REVIEW

**Comment by hemanttanwar** [comment]
fixes #12058

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java [hunk 1]
```diff
@@ -455,21 +419,108 @@ protected ReactorReceiver createConsumer(String entityPath, Receiver receiver,
...
+     * @param linkName Name of the receive link.
+     * @param entityPath Address in the message broker for the link.
+     * @param finalDestinationPath Final destination address in the message broker for the link.
+     * @param linkProperties The properties needed to be set on the link.
+     * @param timeout Operation timeout when creating the link.
+     * @param retry Retry policy to apply when link creation times out.
+     *
+     * @return A new instance of an {@link AmqpLink} with the correct properties set.
+     */
+    protected Mono<AmqpLink> createProducer(String linkName, String entityPath, String finalDestinationPath,
```
Comment (conniey) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java:450]: This is very specific. "final destination"? This belongs in Service Bus.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 2]
```diff
@@ -468,6 +469,20 @@ public ServiceBusSenderClientBuilder queueName(String queueName) {
...
+        /**
+         * Sets the name of the initial destination Service Bus queue to publish messages to.
+         *
+         * @param viaQueueName The initial destination of the message.
+         *
+         * @return The modified {@link ServiceBusSenderClientBuilder} object.
+         *
+         * @see <a href="https://docs.microsoft.com/azure/service-bus-messaging/service-bus-transactions#transfers-and-send-via">Send Via</a>
+         */
+        public ServiceBusSenderClientBuilder viaQueueName(String viaQueueName) {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:481]: Is this only for queues? Does the entity type also apply to subscriptions? If only queues, should we have a runtime check it's not a topic?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 3]
```diff
@@ -23,13 +23,13 @@
...
      *
-     * @param linkName The name of the link.
      * @param entityPath The remote address to connect to for the message broker.
+     * @param viaEntityPath The intermediate remote address to connect to for the message broker.
      * @param retryOptions Options to use when creating the link.
      *
      * @return A new or existing send link that is connected to the given {@code entityPath}.
      */
-    Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, AmqpRetryOptions retryOptions);
+    Mono<AmqpSendLink> createSendLink(String entityPath, String viaEntityPath, AmqpRetryOptions retryOptions);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:32]: It's possible to create multiple send links with to the same entity path. That's why we have a linkName parameter. This shouldn't be repurposed for it.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java (hunk 1 above)
Comment (hemanttanwar) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java]: yes it works only with  queue.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 4]
```diff
@@ -496,6 +511,11 @@ public ServiceBusSenderAsyncClient buildAsyncClient() {
             final MessagingEntityType entityType = validateEntityPaths(logger, connectionStringEntityName, topicName,
                 queueName);
 
+            if (!CoreUtils.isNullOrEmpty(viaQueueName) && !CoreUtils.isNullOrEmpty(topicName)) {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:514]: You should use `entityType` instead of topicName for your comparison

Comment (conniey) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:514]: Also add documentation for this runtime error. It should be an IllegalStateException.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java [hunk 5]
```diff
@@ -551,7 +553,13 @@ public void close() {
 
     private Mono<AmqpSendLink> getSendLink() {
         return connectionProcessor
-            .flatMap(connection -> connection.createSendLink(entityName, entityName, retryOptions))
+            .flatMap(connection -> {
+                String linkName = entityName;
+                if (!CoreUtils.isNullOrEmpty(viaEntityName)) {
+                    linkName = String.format("VIA-%s", viaEntityName);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java:559]: Do we need to do a string format? You can append it.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 6]
```diff
@@ -25,11 +25,13 @@
      *
      * @param linkName The name of the link.
      * @param entityPath The remote address to connect to for the message broker.
+     * @param viaEntityPath The intermediate remote address to connect to for the message broker.
      * @param retryOptions Options to use when creating the link.
      *
      * @return A new or existing send link that is connected to the given {@code entityPath}.
      */
-    Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, AmqpRetryOptions retryOptions);
+    Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, String viaEntityPath,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:33]: viaEntityPath is an optional parameter, I would put this at the end of the parameter list.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java [hunk 7]
```diff
@@ -133,18 +134,22 @@ public ServiceBusReactorAmqpConnection(String connectionId, ConnectionOptions co
...
+        AmqpRetryOptions retryOptions) {
+
+        String entity = !CoreUtils.isNullOrEmpty(viaEntityPath) ? viaEntityPath : entityPath;
+        return createSession(entity).cast(ServiceBusSession.class).flatMap(session -> {
+            logger.verbose("Get or create sender link : '{}'", linkName);
             final AmqpRetryPolicy retryPolicy = RetryUtil.getRetryPolicy(retryOptions);
 
-            return session.createProducer(linkName, entityPath, retryOptions.getTryTimeout(), retryPolicy)
-                .cast(AmqpSendLink.class);
+            return session.createSenderLink(linkName, entityPath, viaEntityPath, retryOptions.getTryTimeout(),
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java:151]: Same with putting viaEntityPath at the end of the parameter list.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java [hunk 8]
```diff
@@ -133,18 +134,22 @@ public ServiceBusReactorAmqpConnection(String connectionId, ConnectionOptions co
...
+     * @param viaEntityPath The intermediate remote address to connect to for the message broker.
      * @param retryOptions Options to use when creating the link.
      *
      * @return A new or existing send link that is connected to the given {@code entityPath}.
      */
     @Override
-    public Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, AmqpRetryOptions retryOptions) {
-        return createSession(entityPath).flatMap(session -> {
-            logger.verbose("Get or create producer for path: '{}'", entityPath);
+    public Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, String viaEntityPath,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java:143]: There are two different names in this file. Is `viaEntityPath` the same as `transferEntityPath`? If so, we should fix the internal implementation to use the same name. This makes it hard to follow in the future.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java [hunk 9]
```diff
@@ -88,6 +98,34 @@
...
+        final Duration serverTimeout = adjustServerTimeout(timeout);
+        Map<Symbol, Object> linkProperties = new HashMap<>();
+
+        linkProperties.put(LINK_TIMEOUT_PROPERTY, UnsignedInteger.valueOf(serverTimeout.toMillis()));
+
+        if (!CoreUtils.isNullOrEmpty(viaEntityPath)) {
+            linkProperties.put(LINK_TRANSFER_DESTINATION_PROPERTY, entityPath);
+            logger.verbose("Get or create sender link for via entity path: '{}'", viaEntityPath);
+
+            final TokenManager tokenManager = tokenManagerProvider.getTokenManager(cbsNodeSupplier, entityPath);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java:117]: The token manager is Closeable. You never dispose of this resource. You don't need the check for endpoint states here the token manager already does it.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSession.java [hunk 10]
```diff
@@ -49,4 +50,19 @@
...
+     * @param linkName Name of the link.
+     * @param entityPath The entity path this link connects to, so that it may send events to the message broker.
+     * @param viaEntityPath The entity path this link connects to, so that it may send events to the message broker
+     *     via this entity.
+     * @param timeout Timeout required for creating and opening an AMQP link.
+     * @param retryPolicy The retry policy to use when consuming messages.
+     *
+     * @return A newly created AMQP link.
+     */
+    Mono<AmqpLink> createSenderLink(String linkName, String entityPath, String viaEntityPath, Duration timeout,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSession.java:66]: put viaEntityPath at the end as it is an optional parameter.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/IntegrationTestBase.java [hunk 11]
```diff
@@ -253,20 +253,32 @@ protected ServiceBusClientBuilder getBuilder(boolean useCredentials) {
 
     protected ServiceBusSenderClientBuilder getSenderBuilder(boolean useCredentials, MessagingEntityType entityType,
         int entityIndex, boolean isSessionAware, boolean sharedConnection) {
+        return getSenderBuilder(useCredentials, entityType, entityIndex, -1, isSessionAware, sharedConnection);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/IntegrationTestBase.java:256]: Do we need to have this in the integration test base? How many test classes actually need the viaQueueName test case? This method returns a builder, you can set that name for the particular test case.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 12]
```diff
@@ -44,6 +46,30 @@
         ENDPOINT, SHARED_ACCESS_KEY_NAME, SHARED_ACCESS_KEY, QUEUE_NAME);
     private static final Proxy PROXY_ADDRESS = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(PROXY_HOST, Integer.parseInt(PROXY_PORT)));
 
+    @Test
+    void viaQueueNameWithTopicNotAllowed() {
+        assertThrows(IllegalArgumentException.class, () -> {
+            final ServiceBusClientBuilder builder = new ServiceBusClientBuilder();
+            builder.connectionString(NAMESPACE_CONNECTION_STRING)
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:53]: This should be scoped to when the build method is actually invoked.

**Comment by conniey** [comment]
Also updating shorthands in your tests (ie. txn -> transaction)

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java [hunk 13]
```diff
@@ -26,41 +26,42 @@
...
 
     /**
      * Creates or gets an existing receive link. The same link is returned if there is an existing receive link with the
      * same {@code linkName}. Otherwise, a new link is created and returned.
      *
      * @param linkName The name of the link.
      * @param entityPath The remote address to connect to for the message broker.
-     * @param transferEntityPath Path if the events should be transferred to another link after being received from this
-     *     link.
+     * @param transferDestinationPath Path if the message should be transferred to another link after being received
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusAmqpConnection.java:42]: Why was this parameter changed? It is an entityPath, similar to the parameter above.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java [hunk 14]
```diff
@@ -134,17 +134,21 @@ public ServiceBusReactorAmqpConnection(String connectionId, ConnectionOptions co
...
      * @return A new or existing send link that is connected to the given {@code entityPath}.
      */
     @Override
-    public Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, AmqpRetryOptions retryOptions) {
-        return createSession(entityPath).flatMap(session -> {
-            logger.verbose("Get or create producer for path: '{}'", entityPath);
+    public Mono<AmqpSendLink> createSendLink(String linkName, String entityPath, AmqpRetryOptions retryOptions,
+         String transferDestinationPath) {
+
+        //String entity = !CoreUtils.isNullOrEmpty(viaEntityPath) ? viaEntityPath : entityPath;
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java:145]: Remove

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java [hunk 15]
```diff
@@ -134,17 +134,21 @@ public ServiceBusReactorAmqpConnection(String connectionId, ConnectionOptions co
...
+        return createSession(entityPath).cast(ServiceBusSession.class).flatMap(session -> {
+            logger.verbose("Get or create sender link : '{}'", linkName);
             final AmqpRetryPolicy retryPolicy = RetryUtil.getRetryPolicy(retryOptions);
 
-            return session.createProducer(linkName, entityPath, retryOptions.getTryTimeout(), retryPolicy)
-                .cast(AmqpSendLink.class);
+            return session.createSenderLink(linkName, entityPath, retryOptions.getTryTimeout(),
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java:150]: The methods in session are all producer and consumer. This adds another name into the mix. Revert back to createProducer

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java [hunk 16]
```diff
@@ -155,12 +159,15 @@ public ServiceBusReactorAmqpConnection(String connectionId, ConnectionOptions co
...
+     * @param transferDestinationPath Path if the events should be transferred to another link after being received
+     *     from this link.
+     * @param entityType {@link MessagingEntityType} to use when creating the link.
      *
      * @return A new or existing receive link that is connected to the given {@code entityPath}.
      */
     @Override
     public Mono<ServiceBusReceiveLink> createReceiveLink(String linkName, String entityPath, ReceiveMode receiveMode,
-        String transferEntityPath, MessagingEntityType entityType) {
+        String transferDestinationPath, MessagingEntityType entityType) {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorAmqpConnection.java:170]: Why the name change? Revert.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java [hunk 17]
```diff
@@ -88,6 +96,34 @@
...
+        if (!CoreUtils.isNullOrEmpty(transferDestinationPath)) {
+            linkProperties.put(LINK_TRANSFER_DESTINATION_PROPERTY, transferDestinationPath);
+            logger.verbose("Get or create sender link {} for via entity path: '{}'", linkName, entityPath);
+
+            final TokenManager tokenManager = tokenManagerProvider.getTokenManager(cbsNodeSupplier,
+                transferDestinationPath);
+
+            return tokenManager.authorize().then(createProducer(linkName, entityPath, timeout, retry,
+                linkProperties))
+                .doFinally(signalType -> tokenManager.close());
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java:120]: We'd be keeping this tokenManager around for the lifetime of this send link when it was only used once. Why not dispose of it after the first authorization?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java [hunk 18]
```diff
@@ -87,11 +88,12 @@
     private final Runnable onClientClose;
     private final String entityName;
     private final ServiceBusConnectionProcessor connectionProcessor;
+    private final String viaEntityName;
 
     /**
      * Creates a new instance of this {@link ServiceBusSenderAsyncClient} that sends messages to a Service Bus entity.
      */
-    ServiceBusSenderAsyncClient(String entityName, MessagingEntityType entityType,
+    ServiceBusSenderAsyncClient(String entityName, String viaEntityName, MessagingEntityType entityType,
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java:96]: This is an optional parameter, we should have this at the end.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java [hunk 19]
```diff
@@ -88,6 +96,38 @@
...
+
+        if (!CoreUtils.isNullOrEmpty(transferDestinationPath)) {
+            linkProperties.put(LINK_TRANSFER_DESTINATION_PROPERTY, transferDestinationPath);
+            logger.verbose("Get or create sender link {} for via entity path: '{}'", linkName, entityPath);
+
+            final TokenManager tokenManager = tokenManagerProvider.getTokenManager(cbsNodeSupplier,
+                transferDestinationPath);
+
+            return tokenManager.authorize()
+                .then(
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSession.java:119]: Why create an entire Mono for this? You can use .doFinally. authorize() is a Mono.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSession.java [hunk 20]
```diff
@@ -49,4 +50,19 @@
...
+     * @param entityPath The entity path this link connects to, so that it may send events to the message broker.
+     * @param timeout Timeout required for creating and opening an AMQP link.
+     * @param retryPolicy The retry policy to use when sending events.
+     * @param transferDestinationPath The entity path this link connects to, so that it may transfer events to
+     *     the message broker via this entity.
+     *
+     * @return A newly created AMQP link.
+     */
+    Mono<AmqpLink> createProducer(String linkName, String entityPath, Duration timeout,
+        AmqpRetryPolicy retryPolicy, String transferDestinationPath);
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSession.java:67]: transferEntityPath

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 21]
```diff
@@ -44,6 +46,33 @@
         ENDPOINT, SHARED_ACCESS_KEY_NAME, SHARED_ACCESS_KEY, QUEUE_NAME);
     private static final Proxy PROXY_ADDRESS = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(PROXY_HOST, Integer.parseInt(PROXY_PORT)));
 
+    @Test
+    void viaQueueNameWithTopicNotAllowed() {
+        // Arrange & Act
+        ServiceBusSenderClientBuilder builder = new ServiceBusClientBuilder()
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:52]: This is just // Arrange. 
The next part is // Act & Assert.

Same with the section below.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 22]
```diff
@@ -44,6 +46,33 @@
...
+            .viaQueueName(VIA_QUEUE_NAME);
+
+        // Assert
+        assertThrows(IllegalStateException.class, () -> builder.buildAsyncClient());
+    }
+
+    @Test
+    void queueClientWithViaQueueName() {
+        // Arrange & Act
+        final ServiceBusSenderAsyncClient client = new ServiceBusClientBuilder()
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:65]: This can be split. // Arrange is where the builder is created.

`// Act` is when you do `var client = builder.buildAsyncClient()`

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSessionTest.java [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+        logger.info("[{}] Tearing down.", testInfo.getDisplayName());
+
+        Mockito.framework().clearInlineMocks();
+    }
+
+    /**
+     * Test for create Sender Link when via-queue is used.
+     */
+    @Test
+    void createViaSenderLink() throws IOException {
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/implementation/ServiceBusReactorSessionTest.java:209]: what about the case where the validation for the token manager fails?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java [hunk 24]
```diff
@@ -551,7 +553,14 @@ public void close() {
...
     private Mono<AmqpSendLink> getSendLink() {
         return connectionProcessor
-            .flatMap(connection -> connection.createSendLink(entityName, entityName, retryOptions))
+            .flatMap(connection -> {
+                if (!CoreUtils.isNullOrEmpty(viaEntityName)) {
+                    return connection.createSendLink("VIA-".concat(viaEntityName), viaEntityName, retryOptions, entityName);
+                } else {
+                    return connection.createSendLink(entityName, entityName, retryOptions, null);
+                }
+
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java:562]: nit: remove new line.
